### PR TITLE
ONNX-IR Refactor - Op/Node-Centric Architecture

### PR DESCRIPTION
**Pull request by antimora**
This PR refactors the ONNX IR layer to be more op/node-centric, significantly simplifying future op additions and maintenance. The previous design made limiting assumptions that didn't scale well as more ops were added.

### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Changes

- Simplified Op Model: Redesigned around individual operations and nodes for better scalability
- Removed TensorData Indirection: Now uses burn-tensor's native tensor data directly, enabling mmap support and better type handling
- Opset-Aware Processing: Processors now accept opset parameters, allowing flexible condition testing across different ONNX versions
- Constants-First Approach: All values treated as constants initially (with initializers and identities converted)
- Explicit Static Input Declaration: Users explicitly declare which inputs are static, rather than preemptive conversion
- Transformation Pipeline: New multi-phase transformation pipeline for better visibility into the processing flow
- Native Type Integration: Direct use of TensorData and Dtype for improved efficiency and consistency. Next we will use mmap to reduce memory overhead.
- Graph input name preservation (sanitized) for easy development and troubleshooting.

### Testing

- new onnx-ir onnx tests that cover the onnx-ir infrastructure and edge cases
- all existing tests pass
- all model-checks pass

### CODE REVIEW

File: contributor-book/src/guides/onnx-to-burn-conversion-tool.md [hunk 1]
```diff
@@ -212,85 +233,213 @@ That's it! The registry automatically generates:
...
+
+- Removes unreferenced constant nodes
+- Constructs the final `OnnxGraph` with inputs, outputs, and nodes
+
+This pipeline ensures that:
+
+- Type information flows correctly through the graph
+- Constants are accessible where needed
+- The graph is optimized (no unnecessary Identity nodes)
+- Only referenced nodes are included in the final output
```
Comment (laggui) [inline contributor-book/src/guides/onnx-to-burn-conversion-tool.md]: These bullet points are essentially a repeat of the phases described just before

File: contributor-book/src/guides/onnx-to-burn-conversion-tool.md [hunk 2]
```diff
@@ -212,85 +233,213 @@ That's it! The registry automatically generates:
...
+- **Separation of concerns**: Configuration extraction is separate from type inference
+- **Error handling**: All methods return `Result` for proper error propagation
+- **Opset awareness**: Methods receive opset version for version-specific behavior
+- **Type preferences**: Nodes can request specific types from their inputs (e.g., requesting Shape
+  type for dynamic operations)
 
-1.  Access the constant values during node configuration
-2.  Process operations like Conv1d that expect weights as direct inputs
-3.  Handle shape-defining inputs needed for operations like Reshape
+### ONNX-IR Modules
```
Comment (laggui) [inline contributor-book/src/guides/onnx-to-burn-conversion-tool.md]: I would remove this section entirely, it doesn't really need explanations with the way it is structured IMO. It is more likely to be out of date if we forget to update it with reflecting changes.

File: contributor-book/src/guides/onnx-to-burn-conversion-tool.md [hunk 3]
```diff
@@ -212,85 +233,213 @@ That's it! The registry automatically generates:
...
+
+- **`proto_conversion.rs`**: Conversion utilities from ONNX protobuf to IR structures
+
+### Burn-Import Modules
 
 The `burn-import` crate is organized into several key modules:
 
+- **`onnx/model_gen.rs`**: Public API (`ModelGen`) for converting ONNX models to Burn code
+
 ### Core Modules
```
Comment (laggui) [inline contributor-book/src/guides/onnx-to-burn-conversion-tool.md]: Same here

File: crates/burn-import/src/burn/graph.rs [hunk 4]
```diff
@@ -515,15 +515,22 @@ impl<PS: PrecisionSettings> BurnGraph<PS> {
...
+        input_types: &[Type],
+        output_types: &[Type],
+    ) {
+        // Handle empty graphs: use provided types directly
+        if self.nodes.is_empty() {
+            // For empty graphs, inputs pass through directly to outputs
+            self.graph_input_types.extend_from_slice(input_types);
+            self.graph_output_types.extend_from_slice(output_types);
+            return;
+        }
```
Comment (laggui) [inline crates/burn-import/src/burn/graph.rs:533]: It looks like input/output types are almost always empty.

Plus, when graphs are "empty" it just means that it doesn't contain any nodes, but it still has at least an input and output, which should be represented by a name too. So similar steps should be logically applied and work by retrieving the inputs / outputs? But right now we only check those from nodes.

Simplest alternative: input should not only be the names, but also include the types or be an enum

File: crates/burn-import/src/burn/node/cast.rs [hunk 5]
```diff
@@ -83,17 +81,11 @@ impl<PS: PrecisionSettings> NodeCodegen<PS> for CastNode {
...
-                    ElementType::Bool => TensorKind::Bool,
-                    ElementType::String => panic!("Cast: String type not supported"),
+                let target_kind = match self.target_dtype {
+                    DType::F32 | DType::F64 | DType::F16 => TensorKind::Float,
+                    DType::I32 | DType::I64 | DType::I8 | DType::U16 | DType::U8 => TensorKind::Int,
+                    DType::Bool => TensorKind::Bool,
+                    _ => panic!("Unsupported DType for Cast: {:?}", self.target_dtype),
                 };
 
                 if input_tensor.kind == target_kind {
```
Comment (laggui) [inline crates/burn-import/src/burn/node/cast.rs:91]: Unrelated to the changes in this PR, but cast should support specific dtypes otherwise this might be incorrect

File: crates/burn-import/src/burn/node/gather.rs [hunk 6]
```diff
@@ -339,10 +224,8 @@ impl<PS: PrecisionSettings> NodeCodegen<PS> for GatherNode {
     }
 
     fn input_types(&self) -> Vec<crate::burn::Type> {
-        match &self.index {
-            GatherIndices::Runtime(index_type) => vec![self.input.clone(), index_type.clone()],
-            GatherIndices::Static(_) => vec![self.input.clone()],
-        }
+        // All indices are now runtime, so always return both input and index
+        vec![self.input.clone(), self.index.clone()]
```
Comment (laggui) [inline crates/burn-import/src/burn/node/gather.rs:228]: Why exactly are they all runtime now?

File: crates/burn-import/src/burn/node/gemm.rs [hunk 7]
```diff
@@ -111,8 +111,17 @@ impl OnnxIntoNode for GemmNode {
...
+        Self::new(
+            a,
+            b,
+            c,
+            output,
+            config.alpha,
+            config.beta,
+            config.trans_a,
+            config.trans_b,
+        )
```
Comment (laggui) [inline crates/burn-import/src/burn/node/gemm.rs:124]: The node should just store the config in a field instead. The same thing goes for all other nodes.

File: crates/burn-import/src/burn/node/linear.rs [hunk 8]
```diff
@@ -131,33 +131,22 @@ impl OnnxIntoNode for LinearNode {
...
-                        Data::Int32s(val) => TensorData::new(val, shape).convert::<f32>(),
-                        Data::Int64s(val) => TensorData::new(val, shape).convert::<f32>(),
-                        _ => panic!("Unsupported tensor element type"),
-                    };
-
-                    Some(tensor_data)
+                ArgType::Tensor(_) | ArgType::Shape(_) | ArgType::Scalar(_) => {
+                    // For Tensor, Shape, and Scalar types, extract the underlying tensor data
+                    // onnx-ir now uses burn_tensor::TensorData directly
+                    Some(value.clone().convert::<f32>())
```
Comment (laggui) [inline crates/burn-import/src/burn/node/linear.rs]: Possible improvement: we should convert to the desired dtype, not strictly f32

File: crates/burn-import/src/burn/node/one_hot.rs [hunk 9]
```diff
@@ -69,8 +67,31 @@ impl OnnxIntoNode for OneHotNode {
...
+            onnx_ir::node::one_hot::OneHotValuesInput::Runtime(_) => {
+                panic!("OneHot with runtime values is not supported in burn-import")
+            }
+        };
+
+        // Derive values_type from output to preserve the correct output element type
+        // Even though onnx-ir stores values as [f32; 2], the output type is determined
+        // by the ONNX model's output element type (e.g., Int64 -> Int)
+        let mut values_type = output.clone();
+        values_type.name = format_ident!("{}_values", output.name);
```
Comment (laggui) [inline crates/burn-import/src/burn/node/one_hot.rs]: `values_type` doesn't appear to be used anymore? So we can probably remove the field.

File: crates/burn-import/src/onnx/model_gen.rs [hunk 10]
```diff
@@ -543,8 +543,12 @@ impl ParsedOnnxGraph {
             .map(|output| output.name.clone())
             .collect();
 
+        // Convert ONNX arguments to Burn types for empty graphs
+        let input_types: Vec<_> = self.0.inputs.iter().map(crate::burn::Type::from).collect();
+        let output_types: Vec<_> = self.0.outputs.iter().map(crate::burn::Type::from).collect();
+
         // Register inputs and outputs with the graph
-        graph.register_input_output(input_names, output_names);
+        graph.register_input_output(input_names, output_names, &input_types, &output_types);
```
Comment (laggui) [inline crates/burn-import/src/onnx/model_gen.rs:552]: Mirroring my comment in `register_input_output`: we should not handle this separately/differently.

File: crates/onnx-ir/README.md [hunk 11]
```diff
@@ -9,15 +9,44 @@ tensor operations.
...
+  - **Graph State** (`graph_state.rs`): Manages mutable state during conversion including node
+    storage, name mappings, and tensor data
+  - **Tensor Store** (`tensor_store.rs`): Centralized storage for tensor data with ID-based access
+
+- **Processing**:
+  - **Processor** (`processor.rs`): Defines the `NodeProcessor` trait with type inference and
+    constant lifting capabilities
+  - **Registry** (`registry.rs`): Centralized registry mapping node types to their processors -
+    **add new node types here**
+  - **Node Implementations** (`node/`): Contains operation-specific processor implementations
```
Comment (laggui) [inline crates/onnx-ir/README.md]: I don't think the detailed code tree is valuable in the README. Users will navigate the tree with the given instructions to add new nodes, and can also refer to the contributing instructions. These types of module docs should stay local to a module IMO (and you currently already have module-level documentation, which is great!).

File: crates/onnx-ir/src/ir/argument.rs [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+pub enum ValueSource {
+    /// Static constant value embedded in the argument (name="" + data_id=Some)
+    Static,
+    /// Points to a constant node output (name="constant1_out1" + data_id=None)
+    Constant,
+    /// Points to a runtime node output (name="conv1_out1" + data_id=None)
+    Dynamic,
+    /// Optional/not provided (name="" + data_id=None)
+    Optional,
+}
```
Comment (laggui) [inline crates/onnx-ir/src/ir/argument.rs:30]: `Some`/`None` here has overlap in semantics with `data_id`. If these should be tied, then we should remove the `data_id` field and change the static variant to `Static(TensorId)`.

File: crates/onnx-ir/src/ir/attribute.rs [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+        } else {
+            panic!("Expected Tensors, got {self:?}");
+        }
+    }
+}
+
+/// Convert AttributeValue to an Argument
+impl From<AttributeValue> for Argument {
+    fn from(attr: AttributeValue) -> Argument {
+        let name = "".to_string();
```
Comment (laggui) [inline crates/onnx-ir/src/ir/attribute.rs]: Should we generate a name instead of using an empty placeholder? iirc some constants were being mishandled with empty names before, but maybe this was fixed

File: crates/onnx-ir/src/ir/graph.rs [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+    /// The inputs of the graph.
+    pub inputs: Vec<Argument>,
+
+    /// The outputs of the graph.
+    pub outputs: Vec<Argument>,
+
+    /// Reference to GraphState to keep tensor data alive for .value() access
+    /// This ensures Arguments can access tensor data via their data_id
+    pub(crate) _graph_data: Option<std::rc::Rc<std::cell::RefCell<crate::graph_state::GraphState>>>,
```
Comment (laggui) [inline crates/onnx-ir/src/ir/graph.rs:23]: The reliance on `_graph_data` here feels like a lifetime hack. Couldn't this have been accomplished with a builder pattern? e.g.
1. Keep `GraphState` only during graph construction in a builder.
2. Consume `GraphState` into an immutable `OnnxGraph` that owns the tensor store.
3. Update `Argument` to hold a reference directly to the tensor store instead of the full `GraphState`.

That way, lifetime is clear, and we don't need to hold onto this graph data just for `.value()` argument access.

File: crates/onnx-ir/src/ir/tensor_data_ext.rs [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+
+    /// Convert to `Vec<i64>`, handling Int32 and Int64 types
+    ///
+    /// Useful for extracting indices, shapes, or other integer arrays that need to be i64.
+    fn to_i64_vec(&self) -> Result<Vec<i64>, burn_tensor::DataError>;
+
+    /// Convert to `Vec<usize>`, handling Int32 and Int64 types
+    ///
+    /// Useful for extracting shape or dimension values.
+    fn to_usize_vec(&self) -> Result<Vec<usize>, burn_tensor::DataError>;
```
Comment (laggui) [inline crates/onnx-ir/src/ir/tensor_data_ext.rs]: This doesn't seem to be used at all?

File: crates/onnx-ir/src/ir/tensor_data_ext.rs [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+            DType::I32 => {
+                let vec_i32 = self.to_vec::<i32>()?;
+                Ok(vec_i32.into_iter().map(|v| v as i64).collect())
+            }
+            other => Err(burn_tensor::DataError::TypeMismatch(format!(
+                "Cannot convert {:?} to Vec<i64>",
+                other
+            ))),
+        }
+    }
```
Comment (laggui) [inline crates/onnx-ir/src/ir/tensor_data_ext.rs:94]: This can be simplified to
```rust
match self.dtype {
    DType::I64 | DType::I32 => self.convert::<i64>().to_vec()?
    other => Err(burn_tensor::DataError::TypeMismatch(format!(
        "Cannot convert {:?} to Vec<i64>",
        other
    ))),
}
```
`convert()` is a no-op if the dtype matches the target.

Same for other vec utilities below

File: crates/onnx-ir/src/node/argmax.rs [hunk 17]
```diff
@@ -1,4 +1,37 @@
...
+//!
+//! - `reduced` (tensor(int64)): Output tensor of indices with datatype int64. Shape depends on `keepdims`:
+//!   - If keepdims=1: same rank as input, reduced dimension becomes size 1
+//!   - If keepdims=0: rank reduced by 1
+//!
+//! ## Opset Versions
+//!
+//! - **Opset 13**: Added bfloat16 to type constraints
+//! - **Opset 12**: Added `select_last_index` attribute
+//! - **Opset 11**: Changed `axis` range to support negative indices [-r, r-1]
```
Comment (laggui) [inline crates/onnx-ir/src/node/argmax.rs:11]: Hmmm I don't think we should duplicate the ONNX spec info here. This can be very error prone, and we can just reference the actual spec via the link.

Listing the supported opset versions for the implementation can be valuable though, so we should keep a short form of that.

File: crates/onnx-ir/src/node/argmax.rs [hunk 18]
```diff
@@ -9,110 +42,138 @@ pub struct ArgMaxConfig {
...
+impl NodeProcessor for ArgMaxProcessor {
+    fn infer_types(
+        &self,
+        node: &mut Node,
+        opset: usize,
+        _output_preferences: &OutputPreferences,
+    ) -> Result<(), ProcessError> {
+        crate::processor::validate_opset(opset, 11)?;
+        crate::processor::validate_input_count(node, 1)?;
+        crate::processor::validate_output_count(node, 1)?;
```
Comment (laggui) [inline crates/onnx-ir/src/node/argmax.rs]: We should make this more straightforward, instead of forcing users to always call functions for validation.

Suggestion: a node spec that defines the supported opset(s), input and output counts (which might actually vary based on opset version!!) and let the node processor handle the checks based on the definitions.

File: crates/onnx-ir/src/node/argmax.rs [hunk 19]
```diff
@@ -9,110 +42,138 @@ pub struct ArgMaxConfig {
...
-        node.name,
-        config.keepdims,
-        node.outputs[0].ty
-    );
+        if axis < 0 {
+            axis += tensor.rank as i64;
+        }
+
+        let config = ArgMaxConfig::new(axis as usize, keepdims);
+        Ok(Some(Box::new(config)))
```
Comment (laggui) [inline crates/onnx-ir/src/node/argmax.rs:156]: This is a bit of a weird pattern, where `extract_config` returns the config just for `convert_nodes` to assign it.

It's sort of a two-step dance: processor returns config, caller assigns it back to node. And this has to be done in order to use the config in other steps. So it's also a bit brittle.

File: crates/onnx-ir/src/phases/node_conversion.rs [hunk 20]
```diff
@@ -0,0 +1,498 @@
...
+
+        let config = processor
+            .extract_config(&node, opset_version)
+            .unwrap_or_else(|e| {
+                panic!(
+                    "Failed to extract config for node {} (type: {:?}): {:?}",
+                    node.name, node.node_type, e
+                )
+            });
+        node.config = config;
```
Comment (laggui) [inline crates/onnx-ir/src/phases/node_conversion.rs:90]: This is brittle. And not actually required by the node conversion, feels more like an artifact of plumbing for the current pipeline.

File: crates/onnx-ir/src/phases/post_processing.rs [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+            .restore_data(tensor_data_clone, next_tensor_id);
+        result
+    };
+
+    // Identity elimination
+    log::debug!("Starting Identity elimination");
+    {
+        let elimination_plan = plan_identity_elimination(&nodes);
+        apply_identity_elimination(&mut nodes, &mut outputs, elimination_plan);
+    }
```
Comment (laggui) [inline crates/onnx-ir/src/phases/post_processing.rs:175]: It's not immediately clear to me why we need to remove identity nodes after the previous steps.

Are these nodes somehow added by previous steps? And if not, why can't we clean this up before?

File: crates/onnx-ir/src/phases/type_inference.rs [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+/// This allows runtime preference collection (e.g., Concat requests Shape after seeing Shape inputs).
+pub(super) fn iterative_type_inference_with_preferences(nodes: &mut [Node], opset: usize) {
+    let registry = get_processor_registry();
+
+    // Track collected preferences: (producer_output_name, consumer_name, pref_type_str)
+    let mut collected_preferences: HashSet<(String, String, String)> = HashSet::new();
+
+    let max_iterations = 100; // Safety limit to prevent infinite loops
+
+    for iteration in 1..=max_iterations {
```
Comment (laggui) [inline crates/onnx-ir/src/phases/type_inference.rs]: This seems overly complex. Is it common to have >2 iterations?

In practice, most ONNX graphs should be feed-forward with clear type propagation, no?

File: crates/onnx-ir/src/phases/type_inference.rs [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+        for i in 0..nodes.len() {
+            // Get preferences for this node
+            let prefs = node_preferences
+                .get(&nodes[i].name)
+                .cloned()
+                .unwrap_or_else(crate::processor::OutputPreferences::new);
+
+            // Run type inference on this node
+            let processor = registry.get(&nodes[i].node_type);
+            let _ = processor.infer_types(&mut nodes[i], opset, &prefs);
```
Comment (laggui) [inline crates/onnx-ir/src/phases/type_inference.rs]: Why are we ignoring the possible errors in `infer_types` result?

File: crates/onnx-ir/src/pipeline.rs [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+    log::debug!(" PHASE 3: Type Inference ");
+    let opset_version = extract_opset_version(model);
+    type_inference::infer_types(&state_rc, opset_version);
+
+    log::debug!(" PHASE 4: Post-processing ");
+    let (mut nodes, inputs, mut outputs) = post_processing::post_process(&state_rc);
+
+    log::debug!(" PHASE 5: Finalization ");
+    finalization::finalize(&mut nodes, inputs, &mut outputs, state_rc)
+}
```
Comment (laggui) [inline crates/onnx-ir/src/pipeline.rs:189]: While the separation is great, each phase here traverses the graph nodes multiple times. The worst one probably being in type inference with an iterative process that can reach up to 100 traversals?

For small graphs, this might be negligible, but could be potentially bad for bigger / more complex models.

File: crates/onnx-ir/src/proto_conversion.rs [hunk 25]
```diff
@@ -1,29 +1,261 @@
...
-use bytemuck::{cast_slice, try_cast_vec};
+use burn_tensor::DType;
+use protobuf::Enum;
 
-fn cast_vec_with_fallback<E: bytemuck::Pod>(raw_data: Vec<u8>) -> Vec<E> {
-    // Zero-copy `try_cast_vec` with fallback when alignment and size are not compatible
-    try_cast_vec(raw_data).unwrap_or_else(|(_e, raw_data)| cast_slice(&raw_data).to_vec())
-}
+/// Minimum required ONNX opset version
+pub const MIN_OPSET_VERSION: i64 = 16;
```
Comment (laggui) [inline crates/onnx-ir/src/proto_conversion.rs]: Why not usize? opset should be unsigned

File: crates/onnx-ir/src/proto_conversion.rs [hunk 26]
```diff
@@ -33,85 +265,103 @@ impl TryFrom<TensorProto> for TensorData {
...
                 }
-                ElementType::Float64 => {
-                    if !tensor.double_data.is_empty() {
-                        Data::Float64s(tensor.double_data)
-                    } else {
-                        Data::Float64s(vec![])
-                    }
+                DType::F32 if expected_elems == 0 => {
+                    // Empty tensor with zero elements
+                    Ok(TensorData::new(Vec::<f32>::new(), shape))
```
Comment (laggui) [inline crates/onnx-ir/src/proto_conversion.rs]: Given that `float_data` is a `Vec<f32>`, both `DType::F32` match arms are equivalent no? Same as empty vec:
```rust
TensorData::new(tensor.float_data, shape)
```
so why do we need to check `if expected_elems == 0`?

File: crates/onnx-ir/src/ir/node.rs [hunk 27]
```diff
@@ -0,0 +1,124 @@
...
+        self.get_config::<T>().unwrap_or_else(|| {
+            panic!(
+                "Node '{}' ({:?}) config is not set or has wrong type. Expected {}",
+                self.name,
+                self.node_type,
+                std::any::type_name::<T>()
+            )
+        })
+    }
+}
```
Comment (laggui) [inline crates/onnx-ir/src/ir/node.rs:96]: 1. This seems to be entirely an artifact of plumbing, not a genuine part of the node. The node processor returns a config via `extract_config`, caller assigns it back to node, just so it can be re-used in later steps like `infer_types` or burn-import conversion.
Ideally, config could be passed to the `infer_types`, but with your current pipeline that iterates through the nodes at each step that might not work. Instead of extracting the config during conversion, which is only useful for the subsequent type inference, type inference should extract the config and pass it along.
For `OnnxIntoNode` (the only other place where the config is retrieved again), the processor can be used to extract it again.

2. The runtime type erasure with added up/down casting requirements can be simplified. At first glance, this seems like a good place to have an associative config type on the `NodeProcessor`. This would remove the need for `NodeConfig` trait entirely.

So with 1 + 2, the node no longer stores a config (remove the field and the two methods), and the type is associated to the node processor implementation for a node.

Comment (antimora) [reply crates/onnx-ir/src/ir/node.rs:96]: We have discussed offline and concluded we should properly redo Graph using Node enums. This will get done in a next PR. The issue is tracked here: https://github.com/tracel-ai/burn/issues/3988

File: crates/onnx-ir/src/phases/node_conversion.rs (hunk 20 above)
Comment (antimora) [reply crates/onnx-ir/src/phases/node_conversion.rs:90]: Agreed. We will redo as a graph of node enums: https://github.com/tracel-ai/burn/issues/3988

File: crates/onnx-ir/src/node/argmax.rs (hunk 19 above)
Comment (antimora) [reply crates/onnx-ir/src/node/argmax.rs:156]: Agreed. We will redo as a graph of node enums: https://github.com/tracel-ai/burn/issues/3988

File: crates/onnx-ir/src/phases/type_inference.rs (hunk 22 above)
Comment (antimora) [reply crates/onnx-ir/src/phases/type_inference.rs]: Yeah you're right. It should not take more 2-3 iterations. I can reduce from 100 to 10. We a warning to catch and re-visit this approach. 

FYI, this iterative approached was the safest and bug free compared to graph re-trigger approach, which would have been more efficient but it was complex.

File: crates/onnx-ir/src/phases/type_inference.rs (hunk 23 above)
Comment (antimora) [reply crates/onnx-ir/src/phases/type_inference.rs]: Good catch! I'll fix it so that the error propagates correctly. I even caught a muted error in my unit tests!

I am also refactoring so that parse_onnx returns results.

File: crates/onnx-ir/src/pipeline.rs (hunk 24 above)
Comment (antimora) [reply crates/onnx-ir/src/pipeline.rs:189]: It used to be much worse and I tried to minimize traversals. There are hard cross node dependencies, such as conversation to Linear. I figured for now the code complex is more important than onnx build conversation optimization. BTW, infer_types should be at most 2-3 times. I also reduced max iteration catch to 10. 

Down the line, this could be optimized even further however the code will be more complex. This simple approach is more preferable because it is more robust against ONNX edge cases. Once we have more tests, we could undertake optmization.

File: crates/onnx-ir/src/phases/post_processing.rs (hunk 21 above)
Comment (antimora) [reply crates/onnx-ir/src/phases/post_processing.rs:175]: Rewiring is easier in the end. Also there are Identity nodes with initializers that act like Constants. BTW, there is a big of refactoring of Identify rewiring in Subgraph because it's recursive.

File: crates/onnx-ir/src/ir/attribute.rs (hunk 13 above)
Comment (antimora) [reply crates/onnx-ir/src/ir/attribute.rs]: Removed the conversion entirely. It wasn't needed for attribute to be converted to argument.

File: crates/onnx-ir/src/ir/graph.rs (hunk 14 above)
Comment (antimora) [reply crates/onnx-ir/src/ir/graph.rs:23]: Understood. I will refactor this along with [Implement memory-efficient ONNX weight loading (lazy/protobuf streaming) ](https://github.com/tracel-ai/burn/issues/3596). This will not remain as graph data. The new new refactored will reference mmapped memory regions.

File: crates/burn-import/src/burn/node/gemm.rs (hunk 7 above)
Comment (antimora) [reply crates/burn-import/src/burn/node/gemm.rs:124]: This was a hold out from the previous design principle: code generator uses its own abstraction layer. I recommend we loosen this restriction and code against the new clean onnx ir graph directly and don't do extra arg transformations, etc. This work should be done after: https://github.com/tracel-ai/burn/issues/3988

Comment (antimora) [reply crates/burn-import/src/burn/node/gemm.rs:124]: I filed this ticket: https://github.com/tracel-ai/burn/issues/3989

File: crates/burn-import/src/onnx/model_gen.rs (hunk 10 above)
Comment (antimora) [reply crates/burn-import/src/onnx/model_gen.rs:552]: This will be refactored as part of https://github.com/tracel-ai/burn/issues/3989

It's a big change if we need to refactored again.

File: crates/burn-import/src/burn/node/gather.rs (hunk 6 above)
Comment (antimora) [reply crates/burn-import/src/burn/node/gather.rs:228]: Keeping the index as a constant arg ( as opposed to static arg) or input. Previously we were lifting them and converting to static but would create a tensor. It's better for performance since select op uses tensor.

File: crates/burn-import/src/burn/graph.rs (hunk 4 above)
Comment (antimora) [reply crates/burn-import/src/burn/graph.rs:533]: I will refactor this as part of https://github.com/tracel-ai/burn/issues/3989. This ended up bigger change for this PR to be left to be removed later.

Comment (laggui) [reply crates/burn-import/src/burn/graph.rs:533]: Sure!

File: crates/onnx-ir/src/phases/type_inference.rs (hunk 22 above)
Comment (laggui) [reply crates/onnx-ir/src/phases/type_inference.rs]: Ok sounds like a more reasonable limit for now :+1:
